### PR TITLE
feat(orderbook): immediate-or-cancel orders

### LINKED ISSUE #622: Limit Order (IOC)
EDIT by @kilrau : we want IOC only for now: https://github.com/ExchangeUnion/xud/issues/622#issuecomment-557157547

I suggest to support FOK & IOC with limit  orders.

Use case:

with getorders I see a peer order that I wish to take.
since I never use market orders :-) I issued a limit order targeting the peer order.

The matching with the peer order failed for some reason (peer order does not exists anymore, swap failed, etc), As a result my take order entered the book and broadcasted to all my peers. This was not my intention.
FOK - fill the order or kill it.
IOC - fill immediate or cancel 

reference - http://optionstradingbeginner.blogspot.com/2009/01/fill-or-kill-fok-immediate-or-cancel.html

### PR DESCRIPTION
**Pull request by sangaman**
Closes #622.

This adds an "immediate-or-cancel" order type that prevents an order from entering the order book after matching is complete. It combines the immediate nature of a market order with the price requirement of a limit order.

### CODE REVIEW

**Comment by kilrau** [comment]
Review by @erkarl

File: lib/cli/placeorder.ts [hunk 1]
```diff
@@ -23,12 +23,19 @@ export const placeOrderBuilder = (argv: Argv, side: OrderSide) => {
...
     describe: 'whether to execute in streaming mode',
+    alias: 's',
     default: false,
   })
   .option('replace_order_id', {
     type: 'string',
+    alias: 'r',
     describe: 'the local order id of a previous order to be replaced',
   })
+  .option('ioc', {
```
Comment (erkarl) [inline lib/cli/placeorder.ts]: We're defining the ioc option here, but this value is never passed to the `PlaceOrderRequest`.

```
➜  xud git:(ioc) ✗ ./bin/xucli buy 0.0001 LTC/BTC 1 --ioc
no matches found
remaining 0.0001 LTC entered the order book as fd3184a0-1ce9-11ea-a368-ebe1579e30d5
➜  xud git:(ioc) ✗ ./bin/xucli orderbook

Trading pair: LTC/BTC
┌───────────────────────────────────────┬───────────────────────────────────────┐
│ Buy                                   │ Sell                                  │
├───────────────────┬───────────────────┼───────────────────┬───────────────────┤
│ Quantity          │ Price             │ Price             │ Quantity          │
├───────────────────┼───────────────────┼───────────────────┼───────────────────┤
│ 0.0001            │ 1                 │                   │                   │
└───────────────────┴───────────────────┴───────────────────┴───────────────────┘
```

Comment (sangaman) [reply lib/cli/placeorder.ts]: Good catch, fixed this.

Comment (erkarl) [reply lib/cli/placeorder.ts]: Thanks. The functionality seems to work now (the order is not entered to the orderbook), but the output from the CLI is misleading:

```
➜  xud git:(ioc) ✗ ./bin/xucli buy 0.0001 LTC/BTC 1 --ioc
no matches found
remaining 0.0001 LTC entered the order book as e7892df0-1cef-11ea-9026-6359673420ce
➜  xud git:(ioc) ✗ ./bin/xucli orderbook

Trading pair: LTC/BTC
┌───────────────────────────────────────┬───────────────────────────────────────┐
│ Buy                                   │ Sell                                  │
├───────────────────┬───────────────────┼───────────────────┬───────────────────┤
│ Quantity          │ Price             │ Price             │ Quantity          │
└───────────────────┴───────────────────┴───────────────────┴───────────────────┘
```

Should we also remove the `remaining X XYZ entered the order book...` for `ioc` orders?

Comment (kilrau) [reply lib/cli/placeorder.ts]: YES WE SHOULD! :+1:

Comment (sangaman) [reply lib/cli/placeorder.ts]: Ahhh good catch, I totally didn't think about how that was misleading. I'll update the cli logic not to display this for `ioc` orders.

Comment (sangaman) [reply lib/cli/placeorder.ts]: This should be fixed now as well.

**Pull request merged**